storage/shard.go

```diff
@@ -330,6 +331,9 @@ func (s *dbShard) tickAndExpire(
 	}
 	start := s.nowFn()
 	s.forEachShardEntry(func(entry *dbShardEntry) bool {
+		if c.IsCancelled() {
+			return false
+		}
```

Comment:
This is already the inner loop, the function passed to `forEachShardEntry` is applied to each batch instead of each entry, which means we are only checking for cancellation every 128 entries.